File: backend/src/main/java/com/realestate/duediligence/service/RiskAssessmentService.java
// backend/src/main/java/com/realestate/duediligence/service/RiskAssessmentService.java
package com.realestate.duediligence.service;

import com.realestate.duediligence.dto.RiskAssessmentResponse;
import com.realestate.duediligence.dto.RiskBreakdownDto;
import com.realestate.duediligence.dto.RiskHistoryDto;

/**
 * Service contract for risk assessment operations.
 *
 * The implementation in RiskAssessmentServiceImpl:
 *   - Delegates computation to RiskScoringEngine
 *   - Handles DB persistence (RiskAssessment + RiskFactor entities)
 *   - Marks old assessments isLatest=false before saving new
 *   - Never re-computes if a fresh assessment already exists (getOrCompute)
 */
public interface RiskAssessmentService {

    /**
     * Returns the latest risk assessment for a property.
     * If none exists, computes one and persists it.
     * If one already exists (isLatest=true), returns it immediately without re-scoring.
     *
     * @param propertyId the property to assess
     * @return summary risk response (no per-factor explanations)
     */
    RiskAssessmentResponse getOrCompute(Long propertyId);

    /**
     * Forces a fresh risk recalculation regardless of existing assessment.
     * Marks the previous latest assessment as isLatest=false.
     * Always calls the scoring engine — no cache bypass.
     *
     * @param propertyId the property to re-score
     * @return fresh summary risk response
     */
    RiskAssessmentResponse recalculate(Long propertyId);

    /**
     * Returns the full per-category breakdown with explanations.
     * If no assessment exists, computes one first (same as getOrCompute).
     *
     * @param propertyId the property to analyze
     * @return breakdown with all 6 factor explanations and recommendations
     */
    RiskBreakdownDto getBreakdown(Long propertyId);

    /**
     * Returns complete assessment history for a property, oldest first.
     * Includes all assessments (not just the latest).
     *
     * @param propertyId the property
     * @return history list with trend metadata
     */
    RiskHistoryDto getHistory(Long propertyId);
}

File: frontend/src/services/riskAssessmentService.js
// frontend/src/services/riskAssessmentService.js
"use client";

import api from "@/services/api";
import { API_ROUTES } from "@/constants/apiRoutes";

/**
 * Risk Assessment API client.
 *
 * All methods return the raw response.data — components use React Query-style
 * hooks (see useRiskAssessment) rather than calling these directly.
 *
 * Errors bubble up as-is — components/hooks handle them.
 */
const riskAssessmentService = {
  /**
   * Fetch or compute risk assessment summary for a property.
   * Fast: returns cached DB value on subsequent calls.
   *
   * @param {number|string} propertyId
   * @returns {Promise<Object>} RiskAssessmentResponse
   */
  async getAssessment(propertyId) {
    const res = await api.get(API_ROUTES.RISK.GET(propertyId));
    return res.data;
  },

  /**
   * Fetch full breakdown with per-factor explanations & recommendations.
   * Heavier: loads all 6 factors from DB (or computes if missing).
   *
   * @param {number|string} propertyId
   * @returns {Promise<Object>} RiskBreakdownDto
   */
  async getBreakdown(propertyId) {
    const res = await api.get(API_ROUTES.RISK.BREAKDOWN(propertyId));
    return res.data;
  },

  /**
   * Fetch complete assessment history (all versions, oldest → newest).
   * Used by trend charts.
   *
   * @param {number|string} propertyId
   * @returns {Promise<Object>} RiskHistoryDto
   */
  async getHistory(propertyId) {
    const res = await api.get(API_ROUTES.RISK.HISTORY(propertyId));
    return res.data;
  },

  /**
   * Force fresh risk recalculation. Previous latest becomes historical.
   * Slow: triggers full PropertyAggregationService call.
   *
   * @param {number|string} propertyId
   * @returns {Promise<Object>} RiskAssessmentResponse
   */
  async recalculate(propertyId) {
    const res = await api.post(API_ROUTES.RISK.RECALCULATE(propertyId));
    return res.data;
  },
};

export default riskAssessmentService;